File: apps/gui/src/models/recent_work.ts
/**
 * Cross-project recent work, exactly as Core's `RecentWorkLoaded` fact
 * publishes it.
 *
 * These mirror the host's `RecentProjectProjection` / `RecentSessionProjection`
 * one-to-one, which in turn mirror Core's `RecentProjectSummary` /
 * `RecentSessionSummary` whitelist DTOs. The frontend must not extend them with
 * a derived title, a transcript path, or a preview: those fields are excluded
 * by the contract on purpose, and a client that invented them would be
 * publishing session content Core deliberately withheld.
 *
 * Core also owns the ordering — sessions by
 * `(last_updated_at DESC, canonical_root ASC, session_id ASC)`, projects
 * aggregated from that already-bounded list — so the GUI renders the arrays in
 * the order they arrive and never re-sorts or re-truncates them.
 */

/**
 * The frontend-contract-v1 capability Core publishes for this inventory. The
 * client names the exact id in its unavailable copy so the gap is checkable
 * against Core's handshake rather than described in prose.
 */
export const RECENT_WORK_CAPABILITY = "runtime.recent_work";

export interface RecentProjectView {
  canonicalRoot: string;
  displayName: string;
  /** Seconds since the epoch, as Core recorded it. */
  lastUpdatedAt: number;
  latestSessionId: string | null;
}

export interface RecentSessionView {
  canonicalRoot: string;
  sessionId: string;
  createdAt: number;
  lastUpdatedAt: number;
  messageCount: number;
  toolCallCount: number;
  commandCount: number;
}

export interface RecentWorkResult {
  outcome: { state: "idle" | "pending" | "confirmed" | "rejected"; reason: string | null };
  projects: RecentProjectView[];
  sessions: RecentSessionView[];
  /** Core's own inventory diagnostics, rendered verbatim. */
  diagnostics: string[];
  pendingCommandId: string | null;
  /** False when Core's handshake published no `runtime.recent_work`. */
  capabilityAvailable: boolean;
}

/**
 * What a screen may render for the Recent section.
 *
 * The three states are distinct on purpose: an absent capability, a Core
 * rejection, and a genuinely empty inventory are different facts, and
 * collapsing them into one empty list would misreport Core.
 */
export type RecentWorkState =
  | { kind: "loading" }
  | { kind: "unavailable"; reason: string }
  | { kind: "failed"; reason: string }
  | { kind: "loaded"; projects: RecentProjectView[]; diagnostics: string[] };

/**
 * Coarse "how long ago" buckets for a Core timestamp.
 *
 * Presentation only: the authority stays the epoch seconds Core published, and
 * the bucket is recomputed on every render rather than stored.
 */
export function relativeAge(
  lastUpdatedAt: number,
  now: number,
): { unit: "now" | "minutes" | "hours" | "days" | "weeks"; count: number } {
  const seconds = Math.max(0, Math.floor(now / 1000) - lastUpdatedAt);
  if (seconds < 60) return { unit: "now", count: 0 };
  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) return { unit: "minutes", count: minutes };
  const hours = Math.floor(minutes / 60);
  if (hours < 24) return { unit: "hours", count: hours };
  const days = Math.floor(hours / 24);
  if (days < 7) return { unit: "days", count: days };
  return { unit: "weeks", count: Math.floor(days / 7) };
}

File: apps/gui/src-tauri/src/recent_work.rs
//! Cross-project recent work, as a client of the Core inventory contract.
//!
//! Core alone scans the shared session home, validates each project key, and
//! bounds the answer (`docs/frontend-integration-contract.md`, "Recent Work
//! Contract"). This module is only the transport-safe vocabulary the webview
//! and the shell share; the typed Core command is built in `adapter.rs`.
//!
//! The GUI never reads a transcript, a project directory, or the session index,
//! and never orders, re-ranks, or truncates the answer: Core already ordered
//! sessions by `(last_updated_at DESC, canonical_root ASC, session_id ASC)` and
//! aggregated projects from that bounded list.

use serde::Serialize;

use crate::d1::D1OutcomeProjection;

/// The frontend-contract-v1 capability that carries the recent-work inventory.
///
/// This is the id Core publishes in its handshake
/// (`FRONTEND_V1_EXTENSION_CAPABILITIES`); the client must not invent a
/// finer-grained one, because an unpublished id can never become available.
pub const RECENT_WORK_CAPABILITY: &str = "runtime.recent_work";

/// One project exactly as `RecentProjectSummary` published it.
///
/// The DTO is a Core whitelist: canonical root, derived display name, latest
/// stable timestamp, latest session id. No transcript path, title, preview, or
/// metadata value exists here to leak, and the client adds none.
#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RecentProjectProjection {
    pub canonical_root: String,
    pub display_name: String,
    pub last_updated_at: u64,
    pub latest_session_id: Option<String>,
}

/// One session exactly as `RecentSessionSummary` published it.
#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RecentSessionProjection {
    pub canonical_root: String,
    pub session_id: String,
    pub created_at: u64,
    pub last_updated_at: u64,
    pub message_count: u64,
    pub tool_call_count: u64,
    pub command_count: u64,
}

/// What the client may render after one `QueryRecentWork`.
///
/// `projects`, `sessions`, and `diagnostics` are populated only from the
/// confirming `RecentWorkLoaded` fact, so a pending or rejected read never
/// leaves the Welcome screen or the project picker showing rows Core did not
/// publish.
#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RecentWorkResult {
    pub outcome: D1OutcomeProjection,
    pub projects: Vec<RecentProjectProjection>,
    pub sessions: Vec<RecentSessionProjection>,
    /// Core's own inventory diagnostics, rendered verbatim by code.
    pub diagnostics: Vec<String>,
    pub pending_command_id: Option<String>,
    pub capability_available: bool,
}
